File: Data Structures & Algorithms/subsets/submission-1.cpp
class Solution {
public:
    vector<vector<int>> retArr;

    void dfs(vector<int>& nums, vector<int>& subset, int index) {
        if (index >= nums.size()) {
            retArr.push_back(subset);
            return;
        }
        subset.push_back(nums[index]);
        dfs(nums, subset, index + 1);
        subset.pop_back();
        dfs(nums, subset, index + 1);
    }

    vector<vector<int>> subsets(vector<int>& nums) {
        vector<int> subset;
        dfs(nums, subset, 0);
        return retArr;
    }
};

File: Data Structures & Algorithms/subsets-ii/submission-1.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
        retArr = []
        subset = []
        nums.sort()

        def dfs(index):
            if index >= len(nums):
                retArr.append(subset.copy())
                return
            subset.append(nums[index])
            dfs(index+1)
            subset.pop()
            while (index+1) < len(nums) and nums[index] == nums[index+1]:
                index += 1
            dfs(index+1)


        dfs(0)
        return retArr